artiq/frontend/aqctl_proxy_moninj.py

```diff
@@ -0,0 +1,280 @@
+        if self.connected:
+            await self.comm.close()
+
+    @property
+    def connected(self):
+        return super().connected and hasattr(self.comm, "_writer")
+
+    # Callbacks
+    async def on_connected(self):
+        if self.connected:
+            return
+
+        await super().on_connected()
+        logging.info("connected to coredev")
+        self._supervisor.notify["connected"]["coredev"] = True
+
+    async def on_disconnected(self):
+        if not self.connected:
+            return
+
+        await super().on_disconnected()
+        logger.error("disconnected from core device")
+        self._supervisor.notify["connected"]["coredev"] = False
+        while self._supervisor.activation_event.is_set() and not self.connected:
+            try:
+                await self.connect(self._supervisor.master.core_addr, 1383)
+            except:
+                logging.error("core device still unreachable, retrying...")
+                await asyncio.sleep(10)
+
+    def on_monitor(self, channel, probe, value):
+        logger.debug(f"received monitor data {(channel, probe, value)}")
+        if channel not in self._supervisor.notify["monitor"].raw_view:
+            self._supervisor.notify["monitor"][channel] = dict()
+        self._supervisor.notify["monitor"][channel][probe] = value
+
+    def on_injection_status(self, channel, override, value):
+        logger.debug(f"received injection status {(channel, override, value)}")
+        if channel not in self._supervisor.notify["injection_status"].raw_view:
+            self._supervisor.notify["injection_status"][channel] = dict()
+        self._supervisor.notify["injection_status"][channel][override] = value
+
+
+class MonInjMaster(Connectable):
+    def __init__(self, supervisor):
+        super().__init__()
+        self._core_addr_cache = None
+        self._supervisor = supervisor
+        self.ddb = dict()
+        self.ddb_notify = Subscriber(notifier_name="devices", notify_cb=self.on_notify,
+                                     target_builder=self.build_ddb, disconnect_cb=self.on_disconnected)
+
+    def build_ddb(self, init):
+        self.ddb = init
+        return self.ddb
+
+    async def connect(self, master, port):
+        logger.debug(f"trying to connect to master server at {master}:{port}")
+        await self.ddb_notify.connect(master, port)
+
+    async def disconnect(self):
+        await self.ddb_notify.close()
+        await self.on_disconnected()
+
+    @property
+    def core_addr(self):
+        return self.ddb["core"]["arguments"]["host"]
+
+    # Callbacks
+    async def on_connected(self):
+        if self.connected:
+            return
+
+        await super().on_connected()
+        logger.info("connected to master")
+        self._supervisor.notify["connected"]["master"] = True
+
+        # if the core device address is not the same of what we memoized, notify the change
+        if hash(self._core_addr_cache) ^ hash(self.core_addr):
```

Comment:
Is there a particular need to use `hash()` here?
As far as I can tell, it only compares the core address.

Reply:
you can't do xor with strings, so i used hash instead

Reply:
Just do string comparison like `!=`?

Reply:
oh yes, i made it overcomplicated